### PR TITLE
Improve shift_jis handling and convert unicode chars to their ascii equivalents

### PR DESCRIPTION
**Pull request by NickCondron**
See https://github.com/project-slippi/slippi-js/blob/master/src/utils/fullwidth.ts

Useful shift jis table: https://uic.io/en/charset/show/cp932/

I added the tilde case, which doesn't exist in slippi-js. I don't know if it's worth keeping.

This conversion is destructive, so we'd have to be more clever if we want to make it reversible. However, for now that is not necessary since we store the whole Game Start event as raw bytes for round-tripping.

Another option is to expose a `melee_shift_jis` type with a `From<melee_shift_jis> for String` for something like that.

### CODE REVIEW

File: peppi/src/serde/shift_jis.rs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+fn fix_char(c: char) -> char {
+	let mut c: u32 = u32::from(c);
+
+	// Performs the following fixes:
+	// 1. Map full-width code points to their half-width equivalents
+	// https://en.wikipedia.org/wiki/Halfwidth_and_Fullwidth_Forms_(Unicode_block)
+	// 2. Change ideographic space (U+3000) to ascii space
+	// 3. Change tilde (called WAVE DASH) (U+301C) to ascii tilde ~
+	// 4. Change Right Single/Double Quotation Mark (U+2019/U+201D) to their
+	// ascii equivalents
```
Comment (NickCondron) [inline peppi/src/serde/shift_jis.rs:21]: This is a summary of the critical addition of this PR

**Comment by hohav** [comment]
This is good functionality, and I agree with adding it to Peppi. And while it's true that this won't currently impact round-tripping, I'm still uneasy about automatic, lossy conversions. Does slippi-js provide any way to access the original values?

**Comment by NickCondron** [comment]
> Does slippi-js provide any way to access the original values?

No

>  I'm still uneasy about automatic, lossy conversions.

Gotcha. I see a potential approach:

```rust
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct MeleeShiftJis(pub(crate) String);

impl MeleeShiftJis {
	pub fn as_str(&self) -> &str {
		self.0.as_str()
	}
	pub fn to_fixed(&self) -> String {
		self.0.clone().chars().map(fix_char).collect::<String>()
	}
}
```

This requires the user (who owns a `MeleeShiftJis` struct) to make a decision on which way they want to access the inner value. Unfortunately it adds a bit of overhead, but I think that's unavoidable if we want to allow access to either form. I could probably be more clever and rewrite `to_fixed` to return a `Cow<'_, str>` to save time in the simple case that no fix is needed.

**Comment by NickCondron** [comment]
I adopted both of your naming suggestions, thanks.

File: peppi/src/model/shift_jis.rs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+		c = c + 0x0020 - 0xff00;
+	} else if c == 0x3000 {
+		c = 0x20;
+	} else if c == 0x301c {
+		c = 0x7e;
+	} else if c == 0x2019 {
+		c = 0x27;
+	} else if c == 0x201d {
+		c = 0x22;
+	}
```
Comment (hohav) [inline peppi/src/model/shift_jis.rs:49]: Could you add comments or use constants to make it clear which of these corresponds to which character or character range?

File: peppi/src/model/shift_jis.rs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+/// 1. Map full-width code points to their half-width equivalents
+/// https://en.wikipedia.org/wiki/Halfwidth_and_Fullwidth_Forms_(Unicode_block)
+/// 2. Change ideographic space (U+3000) to ascii space
+/// 3. Change tilde (called WAVE DASH) (U+301C) to ascii tilde ~
+/// 4. Change Right Single/Double Quotation Mark (U+2019/U+201D) to their
+/// ascii equivalents
+fn fix_char(c: char) -> char {
+	let mut c: u32 = u32::from(c);
+
+	if c > 0xff00 && c < 0xff5f {
```
Comment (hohav) [inline peppi/src/model/shift_jis.rs:39]: It feels weird to represent this as an open range. Did you have a particular reason?

If you use a closed range you could also turn this whole thing into a `match`.

File: peppi/tests/shift_jis.rs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+	let game = game("crazy_name_tags")?;
+	let tag1 = game.start.players[0].name_tag.as_ref().unwrap();
+	let tag2 = game.start.players[1].name_tag.as_ref().unwrap();
+	let tag3 = game.start.players[2].name_tag.as_ref().unwrap();
+	let tag4 = game.start.players[3].name_tag.as_ref().unwrap();
+
+	assert_eq!(tag1.to_normalized(), "! CLOWN");
+	assert_eq!(tag2.to_normalized(), "C@ぞ~");
+	assert_eq!(tag3.to_normalized(), "A $ホ ぬヅ。");
+	assert_eq!(tag4.to_normalized(), "!!!!!!!!");
```
Comment (hohav) [inline peppi/tests/shift_jis.rs:25]: Could you also add assertions about the un-normalized values? Mostly to make the test more self-documenting.

**Comment by NickCondron** [comment]
Thank you for the suggestions! I think the code is a lot tighter now.

**Comment by hohav** [comment]
Really nice work, and sorry I let this sit for so long. Thank you!

**Pull request merged**